### PR TITLE
[static-routing-manager] Add the ability to create routes with "via dev" and without specifying a gateway IP.

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description

Add the ability to create routes with "via dev" and without specifying a gateway IP.
For example `ip route add x.x.x.x/x dev ens19 table vrack`

Such routes can be useful:
- when using ppp connections, 
- in large L2 domains with multiple ip-subnets, 
- to work with routers whose address is unknown in advance.

## Why do we need it, and what problem does it solve?

Cases:
- **The target interface is `Point-to-Point`** (for example PPP).
   It is not necessary to specify the next hop ip-address to transfer traffic over p2p connections.
- **The target interface is a `Multi-Point`** (for example, ethernet).
   The server will think that the "target subnet" is in the same l2 domain with it (right behind the interface), and for each address in this subnet, send ARP to determine the target mac address.
- **The network has a router whose address is not known in advance**. 
   In fact, this is a development of the previous paragraph. For it to work, Proxy ARP must be configured on the router. 
   In this case, the server will try to determine the mac for any address sent through this interface via ARP, and the router will respond with its own mac-address to any request. 
   The negative impact of using this method is an increasing number of arp requests on the network, so you should use it with caution and not configure the route to 0.0.0.0./0 in this way.

Cases 2 and 3 are widely used in popular data centers and clouds.

## Why do we need it in patch release?
It is urgent customer request.

## What is the expected result?

We have the ability to create routes "via dev" without specifying the gateway IP address.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: static-routing-manager
type: feature
summary: Add the ability to create routes with "via dev" and without specifying a gateway IP.
impact: Pods of static-routing-manager will be restarted.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/azure'...

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 1]
```diff
@@ -222,44 +222,171 @@ func periodicalRunEventReconcile(
...
+		Routes: preparedRoutes,
 	}
-	(*rem)[re.getHash()] = re
 }
 
-func (rem *RouteEntryMap) AppendR(route v1alpha1.Route, tbl int) {
+func (rem *RouteEntryMap) AppendRE(re RouteEntry) {
 	if len(*rem) == 0 {
 		*rem = make(map[string]RouteEntry)
 	}
```
Comment (Beastlex) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:389]:
```suggestion
```

**Review by Beastlex — CHANGES_REQUESTED** [review]

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go (hunk 1 above)
Comment (Beastlex) [reply ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:389]: Better to use:
```
func NewIPRuleEntryMap() *IPRuleEntryMap {
	ipRuleEntryMap := make(IPRuleEntryMap)
	return &ipRuleEntryMap
}
g := NewIPRuleEntryMap()
(*g)["something"] = IPRuleEntry{...}
// Use explicit call NewIPRuleEntryMap for initialize in all place
```

Comment (Beastlex) [reply ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:389]: Or sync.Once instead `if len(*rem) == 0` but  it doesn't look good too.

**Comment by Beastlex** [comment]
in this form, you can use the IPRuleEntry type for the key of the map
```
type IPRuleEntry struct {
	Priority   int
	Invert     bool
	Src        string
	Dst        string
	IPProto    int
	SPortRange PortRange
	DPortRange PortRange
	Tos        string
	FWMark     string
	IifName    string
	OifName    string
	UIDRange   UIDRange
	Table      int
} 
m := make(map[IPRuleEntry]bool)
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go (hunk 1 above)
Comment (AndreyPavlovFlant) [reply ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go]: fixed

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/vsphere'...

**Review by Beastlex — APPROVED** [review]

**Review by tarabrind — CHANGES_REQUESTED** [review]

**Review by tarabrind — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/vsphere'...

**Review by z9r5 — APPROVED** [review]

**Review by mortelumina98 — APPROVED** [review]

File: ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml [hunk 2]
```diff
@@ -27,7 +27,10 @@ spec:
                         Целевая подсеть.
                     gateway:
                       description: |
-                        Шлюз, через которую достижима подсеть.
+                        Next Hop — шлюз, через которую достижима подсеть.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml:30]:
```suggestion
                        Шлюз, через которую достижима подсеть.
```

File: ee/modules/025-static-routing-manager/crds/routingtables.yaml [hunk 3]
```diff
@@ -85,7 +85,13 @@ spec:
                       type: string
                       format: ipv4
                       description: |
-                        Nexthop.
+                        Next Hop is the gateway through which a subnet is reachable.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/routingtables.yaml:88]:
```suggestion
                        The gateway through which the subnet is reachable.
```

**Review by mortelumina98 — APPROVED** [review]

File: ee/modules/025-static-routing-manager/crds/internal/sdninternalnoderoutingtables.yaml [hunk 4]
```diff
@@ -91,6 +91,12 @@ spec:
                       format: ipv4
                       description: |
                         Nexthop.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/internal/sdninternalnoderoutingtables.yaml:93]:
```suggestion
                        The gateway through which the subnet is reachable.
```

File: ee/modules/025-static-routing-manager/crds/internal/sdninternalnoderoutingtables.yaml [hunk 5]
```diff
@@ -119,6 +125,12 @@ spec:
...
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/internal/sdninternalnoderoutingtables.yaml:127]:
```suggestion
                        The gateway through which the subnet is reachable.
```

File: ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml [hunk 6]
```diff
@@ -28,6 +28,9 @@ spec:
                     gateway:
                       description: |
                         Шлюз, через которую достижима подсеть.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml:30]:
```suggestion
                        Шлюз, через который достижима подсеть.
```

File: ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml [hunk 7]
```diff
@@ -28,6 +28,9 @@ spec:
...
+                    dev:
+                      description: |
+                        Интерфейс, через который достижима подсеть.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/doc-ru-routingtables.yaml:33]:
```suggestion
                        Имя устройства вывода, через который достижима подсеть.
```

File: ee/modules/025-static-routing-manager/crds/internal/doc-ru-sdninternalnoderoutingtables.yaml [hunk 8]
```diff
@@ -30,3 +30,6 @@ spec:
                     gateway:
                       description: |
                         Шлюз, через который достижима целевая подсеть.
+                    dev:
+                      description: |
+                        Интерфейс, через который достижима подсеть.
```
Comment (mortelumina98) [inline ee/modules/025-static-routing-manager/crds/internal/doc-ru-sdninternalnoderoutingtables.yaml:35]:
```suggestion
                        Имя устройства вывода, через который достижима подсеть.
```

**Review by mortelumina98 — APPROVED** [review]

File: ee/modules/025-static-routing-manager/crds/doc-ru-iprulesets.yaml [hunk 9]
```diff
@@ -18,8 +18,8 @@ spec:
                           Приоритет правила (от него зависит очередность применения).
 
                           При выборе приоритета стоит учитывать несколько нюансов:
-                          - Все правила перебираются сетевой подсистемой linux узла в порядке приоритета (от меньшего к большему) пока не будет найдено совпадение.
-                          - В случае если пакет подпал под какое-то правило и соответствующее действие было успешно выполнено, то перебор правил прекращается.
+                          - Все правила перебираются сетевой подсистемой Linux узла в порядке приоритета (от меньшего к большему) пока не будет найдено совпадение.
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/crds/doc-ru-iprulesets.yaml:21]:
```suggestion
                          - Все правила перебираются сетевой подсистемой Linux-узла в порядке приоритета (от меньшего к большему) пока не будет найдено совпадение.
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 10]
```diff
@@ -222,44 +222,168 @@ func periodicalRunEventReconcile(
...
+		// gwrt, err := netlink.RouteGetWithOptions(net.ParseIP(re.gateway), nil)
+		gwRts, err := netlink.RouteGet(net.ParseIP(preparedRE.gateway))
+		if err != nil || len(gwRts) == 0 {
+			return RouteEntry{}, fmt.Errorf("can not find egress Link by route to Gateway %v, err: %w",
+				route.Gateway,
+				err,
+			)
+		}
+		if len(gwRts) > 1 {
+			fmt.Errorf("more then one egress Link find for Gateway %v. Used only first: %v",
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:350]:
```suggestion
			fmt.Errorf("more then one egress Link found for Gateway %v. Used only first: %v",
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 11]
```diff
@@ -433,7 +581,7 @@ func (nm *nrtMap) updateStateInK8S(ctx context.Context, cl client.Client, log lo
 	for nrtName, ns := range *nm {
 		// Wipe the finalizer if necessary
 		if ns.needToWipeFinalizer && ns.k8sResources.DeletionTimestamp != nil {
-			log.V(config.DebugLvl).Info(fmt.Sprintf("Wipe finalizer on NRT: %v", nrtName))
+			log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Wipe finalizer on NRT: %v", nrtName))
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:584]:
```suggestion
			log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] wipe finalizer on NRT: %v", nrtName))
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 12]
```diff
@@ -450,123 +598,87 @@ func (nm *nrtMap) updateStateInK8S(ctx context.Context, cl client.Client, log lo
...
 			err = cl.Patch(ctx, ns.k8sResources, client.RawPatch(types.MergePatchType, patch))
 			if err != nil {
-				log.Error(err, fmt.Sprintf("unable to patch CR SDNInternalNodeRoutingTable %v, err: %v", nrtName, err))
+				log.Error(err, fmt.Sprintf("[NRTReconciler] unable to patch CR SDNInternalNodeRoutingTable %v, err: %v", nrtName, err))
 			}
 		}
 
 		// Update(patch) status every time
-		log.V(config.DebugLvl).Info(fmt.Sprintf("Update status of NRT: %v", nrtName))
+		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Update status of NRT: %v", nrtName))
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:611]:
```suggestion
		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] update status of NRT: %v", nrtName))
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 13]
```diff
@@ -576,21 +688,21 @@ func delRouteFromNode(route RouteEntry) error {
 // other service functions
 
 func deleteRouteEntriesFromNode(delREM, gdREM RouteEntryMap, actREM *RouteEntryMap, status utils.ReconciliationStatus, log logr.Logger) utils.ReconciliationStatus {
-	for hash, route := range delREM {
-		log.V(config.DebugLvl).Info(fmt.Sprintf("Route %v should be deleted", route))
+	for hash, re := range delREM {
+		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Route %v should be deleted", re))
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:692]:
```suggestion
		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] route %v should be deleted", re))
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 14]
```diff
@@ -599,14 +711,14 @@ func deleteRouteEntriesFromNode(delREM, gdREM RouteEntryMap, actREM *RouteEntryM
...
 func deleteOrphanRoutes(gdREM, actREM RouteEntryMap, log logr.Logger) {
 	log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Starting to find and delete orphan routes (with realm %v) from node.", v1alpha1.D8Realm))
-	for hash, route := range actREM {
+	for hash, re := range actREM {
 		if _, ok := (gdREM)[hash]; ok {
 			continue
 		}
-		log.V(config.DebugLvl).Info(fmt.Sprintf("Route %v should be deleted.", route))
-		err := delRouteFromNode(route)
+		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Route %v should be deleted.", re))
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:718]:
```suggestion
		log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] route %v should be deleted.", re))
```

File: ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go [hunk 15]
```diff
@@ -599,14 +711,14 @@ func deleteRouteEntriesFromNode(delREM, gdREM RouteEntryMap, actREM *RouteEntryM
...
+		err := delRouteFromNode(re)
 		if err != nil {
-			log.V(config.DebugLvl).Info(fmt.Sprintf("Unable to delete route %v,err: %v", route, err))
+			log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] Unable to delete route %v,err: %v", re, err))
```
Comment (apolovov) [inline ee/modules/025-static-routing-manager/images/agent/pkg/controller/routes_reconciler_agent.go:721]:
```suggestion
			log.V(config.DebugLvl).Info(fmt.Sprintf("[NRTReconciler] unable to delete route %v,err: %v", re, err))
```

**Review by apolovov — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/vsphere'...

**Review by apolovov — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [10530](https://github.com/deckhouse/deckhouse/pull/10530) to the branch [release-1.65](https://github.com/deckhouse/deckhouse/tree/release-1.65) successful!

**Pull request merged**